### PR TITLE
Introduce new interactor format

### PR DESCRIPTION
**Pull request by TiteiKo**
ATM, the parameters for the execution are send at the initialization of the instance. This behavior poses 2 main issues:
* it is really hard to make Dependency Injections
* there is no way to reuse an Interactor instance with other parameters

This PR aims at introducing a new format that would solve the previously mentioned issues, all the while still supporting the previous format.

New format example:

```ruby
class Hello
  include Hanami::Interactor
  expose :hello

  def call(target:)
    @hello = "Hello #{target}"
  end

  private

  def valid?(target:)
    !target == 'me'
  end
end

hello = Hello.new
hello.call(target: 'world') # => valid?: true, hello: 'Hello world'
hello.call(target: 'me') # => valid?: false
```

TODO:

- [x] prove that it can work
- [x] write all the current tests for both formats
- [x] write all the documentation

### CODE REVIEW

**Comment by cllns** [comment]
🎉 very excited about this :)

**Comment by gizotti** [comment]
This is awesome. I bumped into the dependency injection problem last week.

**Comment by jodosha** [comment]
@TiteiKo Thanks for taking care of this. 💯  I changed the base branch to `develop` because this should be part of `utils` 1.1.0.

**Comment by AlfonsoUceda** [comment]
@TiteiKo awesome and fantastic work, keep working on this ;)

File: lib/hanami/interactor.rb [hunk 1]
```diff
@@ -476,6 +500,16 @@ def self.extended(interactor)
       end
     end
 
+    def method_added(method_name)
+      return unless method_name == :call
```
Comment (mereghost) [inline lib/hanami/interactor.rb]: isn't it work to bubble the hook up by calling super?

Comment (TiteiKo) [inline lib/hanami/interactor.rb]: Yep indeed, thought of it afterwards, will do

**Comment by mereghost** [comment]
Quick question: does marking the `LegacyInterface` as deprecated be beneficial to us in the long run?

**Comment by TiteiKo** [comment]
@mereghost Yes it would, but I'd like to wait until we have some documentation of the new syntax so we can help people migrate by putting a link in the deprecation.

**Comment by mereghost** [comment]
@hanami/core can we :shipit: ?

File: lib/hanami/interactor.rb [hunk 2]
```diff
@@ -264,6 +263,118 @@ def initialize(*args)
...
+      #   require 'hanami/interactor'
+      #
+      #   class Signup
+      #     include Hanami::Interactor
+      #
+      #     # Method #call is not defined
+      #   end
+      #
+      #   Signup.new.call # => NoMethodError
+      def call(**attrs)
```
Comment (TiteiKo) [inline lib/hanami/interactor.rb]: Could be `call(*args, **attrs)` to really allow any type of params combination... 
```ruby
# will fail with (**attrs) but not with (*args, **attrs)
MyInteractor.new.call('foo', bar: 'baz')
MyInteractor.new.call('foo')
```

WDYT should I make the change?

Comment (cllns) [inline lib/hanami/interactor.rb]: What are the disadvantages? Seems like we should be as flexible as possible, IMO

Comment (mereghost) [inline lib/hanami/interactor.rb]: I think it's more of a matter of what we want to enforce. If we like the more "keyword-ish" way (which I really like, coming from Smalltalk) or just allow anything.

**Comment by TiteiKo** [comment]
@cllns is right, there is no downside and we should be as open as possible. I also agree with @mereghost and prefer kwargs, but maybe we should push this as a best practice and not enforce it...

Anyway, interactors now accept args and kwargs in call :)

Good to merge if good for you

**Comment by TiteiKo** [comment]
@jodosha thanks! :green_heart:

**Pull request merged**